### PR TITLE
RNA Connection: extract connection status into a HOC

### PR DESCRIPTION
**Pull request by sergeymitr**
#### Changes proposed in this Pull Request:
- connection status fetching extracted into HOC `withConnectionStatus`
- Jetpack Backup and Connection UI are changed to utilized that new HOC
- introducing `@wordpress/data` state for RNA Connection package

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
1. Disconnect Jetpack.
2. Go to Jetpack Backup, the connection screen should appear.
3. Click "Set up Jetpack", complete the connection flow.
4. Confirm that Jetpack Backup is connected.
5. Disconnect Jetpack Backup using the "Disconnect" button. After completing it, you should see the "Set up Jetpack" screen.
6. Go to "Tools -> Connection Manager", complete the connection flow. Confirm that it works well.

### CODE REVIEW

**Review by fgiannar — COMMENTED** [review]
Hi @sergeymitr ,

Thanks for working on this! Given I'm not a React expert I hope you won't mind me asking a few more generic questions that will help me better grasp the bigger picture:
- I remember us discussing that we wanted to avoid using Redux. Why do you think this approach will benefit us instead of using React Context + Hooks?
- Seems that Connection package and Backup plugin share the same actions, see `actions/connection-status.js`. Is this ok? 
- I really like the HOC approach in terms of keeping our components DRY. I wonder have you considered hooks (`useContext`) as an alternative here as well?

PS I'm mostly asking to learn not questioning your approach 😄 

Many thanks for your time!

**Comment by sergeymitr** [comment]
Hi @fgiannar, thanks for the review!

**The State**
That wasn't an easy decision to create the package-wide state for RNA Connection.
The main reason why we were against it is keeping the individual components usable separately. They should only rely on properties, and don't load any data from state.
However, it seems that having state is more beneficial that not having it. In this particular case, the state is only being used by the `withConnectionStatus` HOC, so I think there's no harm in having state if we are careful.

There are two goals that we reached by using the state:
1. Making sure that the connection status is only loaded once if multiple instances of `withConnectionStatus` are used.
2. Preventing other instances of `withConnectionStatus` from loading the status if it's already being loaded by another instance.

I decided against using a context and went with the state because the context needs to hold its value somewhere, e.g. in a parent component. I didn't manage to find an elegant solution to make it easy to use while sharing the context value between `withConnectionStatus` instances.
There actually might be a way to overcome that, but I didn't spend too much time on this.
If you have some ideas, we could revisit this 🙂 

**The Actions**
Good catch, the file `actions/connection-status.js` is exactly the same in Connection UI and Backup. The Connection UI was more of an experimental playground, and Backup was built in similar way.
I think it's ok to duplicate the code, since they are entirely separate projects that only happen to live in the same development environment.

**Comment by sergeymitr** [comment]
Hi @gravityrail.
When you get a few minutes, could you take a look at this PR.
I'm particularly interested in your opinion on having a Redux state inside the package. It appears to me that using the state is quite safe, as long as we use it in higher-order components only, and keep the individual "for export" components only accept the state data via properties, and never use the state directly.

File: projects/js-packages/connection/state/store.jsx [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ * Internal dependencies
+ */
+import reducer from './reducers';
+import actions from './actions';
+import selectors from './selectors';
+import storeHolder from './store-holder';
+
+const STORE_ID = 'jetpack-connection';
+
+storeHolder.mayBeInit( STORE_ID, {
```
Comment (gravityrail) [inline projects/js-packages/connection/state/store.jsx:11]: This seems like premature optimization and might cause issues later. Let's start by just initializing this when the package is imported.

**Review by gravityrail — DISMISSED** [review]
This is a really nice improvement. Per our discussion we should merge it and build on top once webpack 5 is merged.

**Review by leogermani — APPROVED** [review]
Looks good for merging

**Comment by sergeymitr** [comment]
Committed to WP.com: rWP230555

**Pull request merged**